src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs

```diff
@@ -98,29 +104,125 @@ public async Task<AuthenticationResult> RunAsync(CancellationToken cancellationT
                     throw;
                 }
                 catch (Exception ex)
                 {
                     apiEvent.ApiErrorCode = ex.GetType().Name;
                     AuthenticationRequestParameters.RequestContext.Logger.ErrorPii(ex);
                     LogMsalErrorTelemetryToClient(ex, telemetryEventDetails, telemetryClients);
+                    ServiceBundle.PlatformProxy.OtelInstrumentation.LogFailedMetrics(
+                        ServiceBundle.PlatformProxy.GetProductName(), 
+                        ex.GetType().Name);
+                    LogMsalFailedTelemetryToOTel(ex);
                     throw;
                 }
                 finally
                 {
+                    ServiceBundle.PlatformProxy.OtelInstrumentation?.StopActivity();
                     telemetryClients.TrackEvent(telemetryEventDetails);
                 }
             }
         }
 
+        private void LogMsalSuccessTelemetryToOTel(AuthenticationResult authenticationResult)
+        {
+            if (ServiceBundle.PlatformProxy.OtelInstrumentation == null)
+            {
+                return;
+            }
+
+            ServiceBundle.PlatformProxy.OtelInstrumentation.LogActivityStatus(true);
+
+            Dictionary<string, object> tags = new Dictionary<string, object> { 
+                { TelemetryConstants.CacheInfoTelemetry, Convert.ToInt64(authenticationResult.AuthenticationResultMetadata.CacheRefreshReason) },
+                { TelemetryConstants.TokenSource, Convert.ToInt64(authenticationResult.AuthenticationResultMetadata.TokenSource) },
+                { TelemetryConstants.Duration, authenticationResult.AuthenticationResultMetadata.DurationTotalInMs },
+                { TelemetryConstants.DurationInCache, authenticationResult.AuthenticationResultMetadata.DurationInCacheInMs },
+                { TelemetryConstants.DurationInHttp, authenticationResult.AuthenticationResultMetadata.DurationInHttpInMs },
+                { TelemetryConstants.TokenType, (int)AuthenticationRequestParameters.RequestContext.ApiEvent.TokenType },
+                { TelemetryConstants.RemainingLifetime, (authenticationResult.ExpiresOn - DateTime.Now).TotalMilliseconds } 
+            };
+
+            if (authenticationResult.AuthenticationResultMetadata.RefreshOn.HasValue)
+            {
+                tags.Add(TelemetryConstants.RefreshOn, DateTimeHelpers.DateTimeToUnixTimestampMilliseconds(authenticationResult.AuthenticationResultMetadata.RefreshOn.Value));
+            }
+            tags.Add(TelemetryConstants.AssertionType, (int)AuthenticationRequestParameters.RequestContext.ApiEvent.AssertionType);
+            tags.Add(TelemetryConstants.Endpoint, AuthenticationRequestParameters.Authority.AuthorityInfo.CanonicalAuthority.ToString());
+            tags.Add(TelemetryConstants.CacheLevel, (int)authenticationResult.AuthenticationResultMetadata.CacheLevel);
+
+            var resourceAndScopes = ParseScopesForTelemetry();
```

Comment:
I think this can be deconstructed something like `(string resouce, string scopes) = ParseScopesForTelemetry();` for better readability.